### PR TITLE
Assignment7+8 linq: @GeorgeKoralski007 & @Crawfordkid2

### PR DESCRIPTION
**Pull request by GeorgeKoralski007**
Finished 1-7 with tests

### CODE REVIEW

**Review by Bhopss152 — COMMENTED** [review]
1.	Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
o	Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
o	Using LINQ, skip the first row in the People.csv. ✔
o	Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
2.	Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
o	Use ISampleData.CsvRows for your data source. ✔
o	Don't forget the list should be unique. ✔
o	Sort the list alphabetically. ✔
o	Include a test that leverages a hardcoded list of addresses. ✔
o	Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
3.	Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
o	Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
o	Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
o	Given the array, consider using string.Join to combine the list into a single string. ✔
4.	Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
o	Use ISampleData.CsvRows as the source of the data. ✔
o	Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
o	Be sure that Person.Address is also populated. ✔
o	Adding null validation to all the Person and Address properties is optional.
o	Consider using ISampleData.CsvRows in your test to verify your results. ✔
5.	Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
o	Use ISampleData.People for your data source. ✔
6.	Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
o	Use the people parameter from ISampleData.People property for your data source. ✔
o	At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
o	Don't forget the list should be unique. ✔
o	It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result. ❌
7.	Given the implementation of Node in Assignment5
•	_**Implement IEnumerable<T> to return all the items in the "circle" of items. ❌
•	Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. ❌

Looks good I think your missing #7 on the list id double check that one.**_ 

Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔

**_and enabled .NET analyzers for both projects ❌_**

**_For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌_**

All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by BenagainB** [comment]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
o Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
o Using LINQ, skip the first row in the People.csv. ✔
o Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
o Use ISampleData.CsvRows for your data source. ✔
o Don't forget the list should be unique. ✔
o Sort the list alphabetically. ✔
o Include a test that leverages a hardcoded list of addresses. ✔
o Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
o Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
o Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
o Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
o Use ISampleData.CsvRows as the source of the data. ✔
o Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
o Be sure that Person.Address is also populated. ✔
o Adding null validation to all the Person and Address properties is optional.
o Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate filter) to return a list of names where the email address matches the filter. ✔
o Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable people) to return a string that contains a unique, comma-separated list of states. ✔
o Use the people parameter from ISampleData.People property for your data source. ✔
o At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
o Don't forget the list should be unique. ✔
o It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result. ✔
Given the implementation of Node in Assignment5
• _**Implement IEnumerable to return all the items in the "circle" of items. 
• Add an IEnumberable ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. 


Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
You've got the Directory.Build>targets file, just need to make sure to drop in properties and items.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔

and enabled .NET analyzers for both projects ✔

For this assignment, consider using Assert.AreEqual() (the generic version) ✔

All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall, project is looking good. Revisit Directory.Build.Props.

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,30 +1,45 @@
...
+using System.Linq;
 
 namespace Assignment
 {
     public class SampleData : ISampleData
     {
+        private IEnumerable<string> rows;
+
+        public SampleData() {
+            this.rows = File.ReadAllLines("People.csv").Skip(1);
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:15]: You could probably just do this on line 19 and remove this constructor and property

**Review by mmwoodfo — COMMENTED** [review]
Good job!

**Pull request closed without merging**